src/SignalR/common/Protocols.MessagePack/src/Protocol/MessagePackHubProtocol.cs

```diff
@@ -358,247 +358,236 @@ private static object[] BindArguments(byte[] input, ref int offset, IReadOnlyLis
+                    WriteCancelInvocationMessage(cancelInvocationMessage, ref writer);
                     break;
                 case PingMessage pingMessage:
-                    WritePingMessage(pingMessage, packer);
+                    WritePingMessage(pingMessage, ref writer);
                     break;
                 case CloseMessage closeMessage:
-                    WriteCloseMessage(closeMessage, packer);
+                    WriteCloseMessage(closeMessage, ref writer);
                     break;
                 default:
                     throw new InvalidDataException($"Unexpected message type: {message.GetType().Name}");
             }
         }
 
-        private void WriteInvocationMessage(InvocationMessage message, Stream packer)
+        private void WriteInvocationMessage(InvocationMessage message, ref MessagePackWriter writer)
         {
-            MessagePackBinary.WriteArrayHeader(packer, 6);
+            writer.WriteArrayHeader(6);
 
-            MessagePackBinary.WriteInt32(packer, HubProtocolConstants.InvocationMessageType);
-            PackHeaders(packer, message.Headers);
+            writer.WriteInt32(HubProtocolConstants.InvocationMessageType);
+            PackHeaders(message.Headers, ref writer);
             if (string.IsNullOrEmpty(message.InvocationId))
             {
-                MessagePackBinary.WriteNil(packer);
+                writer.WriteNil();
             }
             else
             {
-                MessagePackBinary.WriteString(packer, message.InvocationId);
+                writer.Write(message.InvocationId);
             }
-            MessagePackBinary.WriteString(packer, message.Target);
-            MessagePackBinary.WriteArrayHeader(packer, message.Arguments.Length);
+            writer.Write(message.Target);
+            writer.WriteArrayHeader(message.Arguments.Length);
             foreach (var arg in message.Arguments)
             {
-                WriteArgument(arg, packer);
+                WriteArgument(arg, ref writer);
             }
 
-            WriteStreamIds(message.StreamIds, packer);
+            WriteStreamIds(message.StreamIds, ref writer);
         }
 
-        private void WriteStreamInvocationMessage(StreamInvocationMessage message, Stream packer)
+        private void WriteStreamInvocationMessage(StreamInvocationMessage message, ref MessagePackWriter writer)
         {
-            MessagePackBinary.WriteArrayHeader(packer, 6);
+            writer.WriteArrayHeader(6);
 
-            MessagePackBinary.WriteInt16(packer, HubProtocolConstants.StreamInvocationMessageType);
-            PackHeaders(packer, message.Headers);
-            MessagePackBinary.WriteString(packer, message.InvocationId);
-            MessagePackBinary.WriteString(packer, message.Target);
+            writer.WriteInt16(HubProtocolConstants.StreamInvocationMessageType);
+            PackHeaders(message.Headers, ref writer);
+            writer.Write(message.InvocationId);
```

Comment:
I seem to remember there being an oddity with writing strings when I played with v2 of the library a few months ago. We'll need to see if that's been changed or if this causes issues.

Reply:
I saw that too, but could not fully be sure of the underlying impact, here is what i observed to far :
* the `writer` now have a direct `public void Write(string value);`
* there is now 2 signature for `WriteString` :
  - `public void WriteString(in ReadOnlySequence<byte> utf8stringBytes);`
    Writes out an array of bytes that (may) represent a UTF-8 encoded string, prefixed with the length using one of these message codes:
    `MessagePack.MessagePackCode.MinFixStr, MessagePack.MessagePackCode.Str8, MessagePack.MessagePackCode.Str16, or MessagePack.MessagePackCode.Str32.`
  - `public void WriteString(ReadOnlySpan<byte> utf8stringBytes);`
    Writes out an array of bytes that (may) represent a UTF-8 encoded string, prefixed with the length using one of these message codes: 
    `MessagePack.MessagePackCode.MinFixStr, MessagePack.MessagePackCode.Str8, MessagePack.MessagePackCode.Str16, or MessagePack.MessagePackCode.Str32.`
* `System.String` as an `AsSpan()` returning a `ReadOnlySpan<char>`

At first i wondered if i could use `writer.WriteString(foo.AsSpan())` <== but i'm not sure how to properly cast `ReadOnlySpan<char>` to `ReadOnlySpan<byte>`.
Also since there is a `writer.Write(string value)` <== for now i assumed that this API of the `MessagePackWriter` was properly taking care of the string encoding, but as you said this will need to be asserted.
Also i wonder if a client using `MessagePack v2` using the new `Write(string value)` is encoded the same way as the old `MessagePack v1` using `WriteString(string value)`, especially ... will it break deserialization

Reply:
The natural substitute for `MessagePackBinary.WriteString` is `MessagePackWriter.Write(string)`. No need to avoid that.

The oddity that @BrennanConroy may remember may be that at one point we did not allow `null` strings in that method. But we do now. Also we had a bug in deserializing strings that spanned multiple chunks in memory but that too is now fixed.

Reply:
Do you know if the 'oddity' about `null string` need a code change somehow ?
(I guess without doing a breaking change in the API or the protocol itself)